DESIGN.md

```diff
@@ -73,77 +157,225 @@ Flow control support:
-- sync, cold, pull (ie. iterable, file, range)
-- async, cold, pull (ie. RPC/REST network call, cross-thread queue draining)
- 
-Unknown:
+##### Operator
 
-- hot, pull (what is an example of this?)
+An operator follows a specific lifecycle (union of the producer/consumer contract).
 
-Flow control support:
+- It must propagate the `subscribe` event upstream (to the producer).
+- It must obey the RxJava contract (serialize all events, `onError`/`onComplete` are terminal).
+- If it has resources to cleanup it is responsible for watching `onError`, `onComplete`, and `cancel/dispose`, and doing the necessary cleanup.
+- It must propagate the `cancel/dispose` upstream.
+
+In the addition of the previous rules, an operator for `Flowable`:
+
+- It must propagate/negotiate the `request(n)` event.
+
+
+### Creation
+
+Creation of the various types should be exposed through factory methods that provide safe construction.
+
+```java
+Flowable.create(SyncGenerator generator)
+
+Flowable.create(AsyncGenerator generator)
+
+Observable<T>.create(OnSubscribe<Observer<T>> onSubscribe)
+
+Single<T>.create(OnSubscribe<Single.Subscriber<T>> onSubscribe)
+
+Completable<T>.create(OnSubscribe<Completable.Subscriber<T>> onSubscribe)
+```
+
+### Terminal behavior
+
+A producer can terminate a stream by emitting `onComplete` or `onError`. A consumer can terminate a stream by calling `cancel`/`dispose`.
+
+Any resource cleanup of the source or operators must account for any of these three termination events. In other words, if an operator needs cleanup, then it should register the cleanup callback with `cancel`/`dispose`, `onError` and `onComplete`.
+
+The final `subscribe` will *not* invoke `cancel`/`dispose` after receiving an `onComplete` or `onError`.
+
+### JVM target and source compatibility
+
+The 2.x version will target JDK6+ but the source will be using some of the Java 8 features.
+We'll use the retrolambda project to generate JDK6 Byte-code from the Java 8 sources.
+The only Java 8 features we intend to use are: lambda expressions, method references and try-with-resources statements.
+Default methods may be considered if retrolambda support reach an acceptable level of maturity.
+
+It still up to discussion to know how we will generate multiple targets?
+Options are:
+
+- two artifcat ids: rxjava-jdk6, rxjava-jdk8
+- classifiers: jdk6, jdk8
 
-- If `request(n)` behavior is supported in the stream implementation, then:
- - pull-based creation must support `request(n)` semantics
- - push-based creation must provide a default *onBackpressure* strategy
-- If `request(n)` behavior is not supported in the stream implementation, then:
- - push-based creation can push without consideration of a backpressure strategy
- - pull-based creation should be discouraged
+The intent is to let Android users consume the new version of RxJava.
+(Not sure if JDK6 is the appropriate requirement, JDK7+ is maybe sufficient).
```

Comment:
No one will be compiling RxJava sources with Jack so I'm not sure how
that's relevant.

On Sat, Mar 19, 2016, 4:48 PM Jonathan <notifications@github.com> wrote:

> In DESIGN.md
> <https://github.com/ReactiveX/RxJava/pull/3782#discussion_r56754911>:
>
> >
> > -- If `request(n)` behavior is supported in the stream implementation, then:
> > - - pull-based creation must support `request(n)` semantics
> > - - push-based creation must provide a default *onBackpressure* strategy
> > -- If `request(n)` behavior is not supported in the stream implementation, then:
> > - - push-based creation can push without consideration of a backpressure strategy
> > - - pull-based creation should be discouraged
> > +The intent is to let Android users consume the new version of RxJava.
> > +(Not sure if JDK6 is the appropriate requirement, JDK7+ is maybe sufficient).
>
> Also, "The Android N bases its implementation of lambda expressions on
> anonymous classes. This approach allows them to be backwards compatible and
> executable on earlier versions of Android." I think Retrolambda isn't
> compatible with Android's Jack tool.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/ReactiveX/RxJava/pull/3782/files/c1180b2f0cb79bef3ba71c20fbdc061359bf5f9d#r56754911>
>